nncf/openvino/quantization/weights_compression.py

```diff
@@ -9,23 +9,326 @@
+    val = np.max(layer_err)
+    return val
+
+
+def _calculate_scale_per_group(
+    weight: np.array, reduction_axes: Union[int, Tuple[int]], group_size: int = -1
+) -> Tuple[np.array, np.array]:
+    """
+    Calculates scale and reshapes weights for group-wise quantization.
+    Having weights with shapes [c_out, c_in] and group size = 128, the shape of scale is [c_out, c_in // 128, 1], and
+    shape of weights is [c_out, c_in // 128, 128].
+
+    :param weight: Weight array to compress.
+    :param reduction_axes: Axis or axes along which to reduce (collect) different statistics (e.g. min, max).
+    :param group_size: number of weights (e.g. 64 or 128) that are independently quantized or share compression
+        parameters (scale). When it equals -1, per-channel quantization is assumed with group size equals to the size of
+        channel. Defaults to -1.
+    :return: Scale and reshaped weights.
+    """
+    if isinstance(reduction_axes, tuple) and len(reduction_axes) != 1:
+        raise RuntimeError(
+            f"group-quantization is supported for a single reduction axes, but got {len(reduction_axes)}"
+        )
+    reduction_axis = reduction_axes[0] if isinstance(reduction_axes, tuple) else reduction_axes
+    channel_size = weight.shape[reduction_axis]
+    if channel_size % group_size != 0:
+        raise RuntimeError(f"Channel size {channel_size} should be divisible by size of group {group_size}")
+
+    num_groups_per_channel = channel_size // group_size
+    shape = list(weight.shape)  # [a1, r, a2] - "r" refers to number of channels along reduction axis
+    shape[reduction_axis : reduction_axis + 1] = (num_groups_per_channel, group_size)
+    reshaped_weight = weight.reshape(shape)  # [a1, r, a2] -> [a1, r//gs, gs, a2], when "gs" is group size
+    scale = np.max(np.abs(reshaped_weight), axis=reduction_axis + 1, keepdims=True)  # [a1, r//gs, 1, a2]
+    return scale, reshaped_weight
+
+
+def _get_norm_weight_and_nf4_scale(
+    weight: np.array, reduction_axes: Tuple[int], group_size: int = -1
+) -> Tuple[np.array, np.array]:
+    """
+    Calculates scale for nf4 quantization and normalizes weights by the scale.
+    Weights are reshaped in case of positive value of group size.
+
+    :param weight: Weight array to compress.
+    :param reduction_axes: Axis or axes along which to reduce (collect) different statistics (e.g. min, max).
+    :param group_size: number of weights (e.g. 64 or 128) that are independently quantized or share compression
+        parameters (scale). When it equals -1, per-channel quantization is assumed with group size equals to the size of
+        channel. Defaults to -1
+    :return: Normalized weights and nf4 scale.
+    """
+    if group_size != -1:
+        scale, weight = _calculate_scale_per_group(weight, reduction_axes, group_size)  # [a1, r//gs, 1, a2]
+    else:
+        scale = np.max(np.abs(weight), axis=reduction_axes, keepdims=True)  # [a1, 1, a2]
```

Comment:
Should it be 1 or gs over here?
```suggestion
        scale, weight = _calculate_scale_per_group(weight, reduction_axes, group_size)  # [a1, r//gs, gs, a2]
    else:
        scale = np.max(np.abs(weight), axis=reduction_axes, keepdims=True)  # [a1, r, a2]
```

Reply:
The comment has a correct shape for scale. There should be 1 along the reduction axis.